Support Elixir 1.18

scohen opened the pull request:
Support 1.18. 

Test failures were due to old versions of beam files hanging around. This quiets warnings and all tests pass locally.

Fixes #846

scohen committed:
[feat] Attempt to support elixir 1.18

scohen committed:
elixir 1.18 doesn't support otp 25

scohen committed:
fixed compile warning

scohen committed:
[1.18] Updated deps to latest to fix warnings

scohen committed:
[1.18] Add 1.18.1 to test matrix

scohen committed:
[1.18] Quieted compilation warnings

scohen committed:
Update Nix hash of Mix deps

scohen committed:
[fix] Supported pre-1.14 slicing function

scohen committed:
Downgraded elixir_sense

scohen committed:
Update Nix hash of Mix deps

scohen committed:
Updatd all deps

scohen committed:
Update Nix hash of Mix deps

Moosieus commented:
I keep seeing errors pop up, seemingly due to some issue with elixir_sense. Logs [here](https://gist.github.com/Moosieus/7447df68633b442117b804753e1609d0)

scohen committed:
Updated elixir_sense

scohen commented:
@Moosieus Updated elixir_sense. That should take care of the error you're seeing (I was told it's benign, but yes, it's gross).

scohen committed:
Update Nix hash of Mix deps

Moosieus commented:
> I was told it's benign, but yes, it's gross

That about sums up my experience. I've pulled the update, gonna work with it and see if anything else crops up.

Moosieus commented:
That seems to have done the trick

Moosieus reviewed (approved):
Been using without issues, and the code looks good on review.

scohen committed:
disabled tests on versions not supported by elixir sense

DnOberon commented:
Is this kinda dead in the water? I know they're working on the unified LS - but is lexical still receiving updates in the meantime?

rdlu commented:
> Is this kinda dead in the water? I know they're working on the unified LS - but is lexical still receiving updates in the meantime?

I was waiting this pr to be merged as well, but then I installed manually using the 1.18 branch. I went back to next-ls, but this is much faster on my setup.

Cloning, switching to `support_1_18` branch and then `mix package --path ~/.local/bin/lexical` was less painful than I imagined.

It works with my lazyvim/neovim setup: https://github.com/rdlu/dotfiles/blob/main/lazyvim/.config/nvim/lua/plugins/elixir.lua

Also with zed editor I just symlinked to `~/.local/share/zed/extensions/work/elixir/lexical-v0.7.x`

garlandcrow commented:
> > Is this kinda dead in the water? I know they're working on the unified LS - but is lexical still receiving updates in the meantime?
> 
> I was waiting this pr to be merged as well, but then I installed manually using the 1.18 branch. I went back to next-ls for a while, but this is much faster on my setup.
> 
> Cloning, switching to `support_1_18` branch and then `mix package --path ~/.local/bin/lexical` was less painful than I imagined.
> 
> It works with my lazyvim/neovim setup: https://github.com/rdlu/dotfiles/blob/main/lazyvim/.config/nvim/lua/plugins/elixir.lua
> 
> Also with zed editor I just symlinked to `~/.local/share/zed/extensions/work/elixir/lexical-v0.7.x`

hmm I tried you config (Im also using lazyvim) and there wasnt really any difference than what I had before except that path. Built from that branch and I got the following so maybe still some issues??
```
22:19:47.118 [debug] Child {LoggerFileBackend, :general_log} of Supervisor Logger.Backends.Supervisor started
Pid: #PID<0.113.0>
Start Call: Logger.Backends.Watcher.start_link({{LoggerFileBackend, :general_log}, {LoggerFileBackend, :general_log}})
Restart: :transient
Shutdown: 5000
Type: :worker
22:19:47.118 [info] Child Logger.Backends.Supervisor of Supervisor Logger.Backends.Internal started
Pid: #PID<0.112.0>
Start Call: Logger.Backends.Supervisor.start_link([{LoggerFileBackend, :general_log}])
Restart: :permanent
Shutdown: :infinity
Type: :supervisor
22:19:47.118 [info] Child Logger.Backends.Internal of Supervisor Logger.Supervisor started
Pid: #PID<0.108.0>
Start Call: Logger.Backends.Internal.start_link([])
Restart: :permanent
Shutdown: :infinity
Type: :supervisor
22:19:47.118 [info] Application logger started at :nonode@nohost
22:19:48.017 [info] Child :ttb_autostart of Supervisor :runtime_tools_sup started
Pid: #PID<0.1538.0>
Start Call: :ttb_autostart.start_link()
Restart: :temporary
Shutdown: 3000
Type: :worker
22:19:48.017 [info] Application runtime_tools started at :nonode@nohost
22:19:48.018 [info] Application erts started at :nonode@nohost
22:19:48.018 [info] Application lx_lexical_shared started at :nonode@nohost
22:19:48.018 [info] Application lx_lexical_test started at :nonode@nohost
22:19:48.018 [info] Child LXSnowflake.Generator of Supervisor #PID<0.1541.0> (Supervisor.Default) started
Pid: #PID<0.1542.0>
Start Call: LXSnowflake.Generator.start_link(1704070800000, 1)
Restart: :permanent
Shutdown: 5000
Type: :worker
22:19:48.018 [info] Application lx_snowflake started at :nonode@nohost
22:19:48.019 [info] Application lx_sourceror started at :nonode@nohost
22:19:48.019 [info] Application lx_common started at :nonode@nohost
22:19:48.019 [info] Application lx_elixir_sense started at :nonode@nohost
22:19:48.020 [info] Application jason started at :nonode@nohost
22:19:48.020 [info] Application logger_file_backend started at :nonode@nohost
22:19:48.020 [info] Application lx_path_glob started at :nonode@nohost
22:19:48.020 [info] Application lx_proto started at :nonode@nohost
22:19:48.020 [info] Application lx_protocol started at :nonode@nohost
22:19:48.028 [info] Child LXical.Document.Store of Supervisor LXical.Server.Supervisor started
Pid: #PID<0.1546.0>
Start Call: LXical.Document.Store.start_link([derive: [analysis: &LXical.Ast.analyze/1]])
Restart: :permanent
Shutdown: 5000
Type: :worker
22:19:48.030 [info] Child LXical.Server of Supervisor LXical.Server.Supervisor started
Pid: #PID<0.1547.0>
Start Call: LXical.Server.start_link([])
Restart: :permanent
Shutdown: 5000
Type: :worker
22:19:48.030 [info] Child LXical.Server.ProjectSupervisor of Supervisor LXical.Server.Supervisor started
Pid: #PID<0.1548.0>
Start Call: DynamicSupervisor.start_link([name: LXical.Server.ProjectSupervisor, strategy: :one_for_one])
Restart: :permanent
Shutdown: :infinity
Type: :supervisor
22:19:48.030 [info] Child LXical.Server.TaskQueue.State.TaskSupervisor of Supervisor LXical.Server.Supervisor started
Pid: #PID<0.1549.0>
Start Call: Task.Supervisor.start_link([name: LXical.Server.TaskQueue.State.TaskSupervisor])
Restart: :permanent
Shutdown: :infinity
Type: :supervisor
22:19:48.030 [info] Child LXical.Server.TaskQueue of Supervisor LXical.Server.Supervisor started
Pid: #PID<0.1550.0>
Start Call: LXical.Server.TaskQueue.start_link([])
Restart: :permanent
Shutdown: 5000
Type: :worker
22:19:48.030 [info] Child LXical.Server.Transport.StdIO of Supervisor LXical.Server.Supervisor started
Pid: #PID<0.1551.0>
Start Call: LXical.Server.Transport.StdIO.start_link(:standard_io, &LXical.Server.protocol_message/1)
Restart: :permanent
Shutdown: 5000
Type: :worker
22:19:48.030 [info] Application lx_server started at :nonode@nohost
22:19:48.041 [info] Starting project at uri file:///Users/smithy/Repos/test
22:19:48.051 [info] Child {LXical.Server.Project.Progress, "test"} of Supervisor :"test::supervisor" started
Pid: #PID<0.1553.0>
Start Call: LXical.Server.Project.Progress.start_link(%LXical.Project{root_uri: "file:///Users/smithy/Repos/test", mix_exs_uri: "file:///Users/smithy/Repos/test/mix.exs", mix_project?: true, mix_env: nil, mix_target: nil, env_variables: %{}, project_module: nil, entropy: 11263})
Restart: :permanent
Shutdown: 5000
Type: :worker
22:19:48.053 [info] Child {LXical.RemoteControl.ProjectNodeSupervisor, "test"} of Supervisor :"test::supervisor" started
Pid: #PID<0.1554.0>
Start Call: LXical.RemoteControl.ProjectNodeSupervisor.start_link(%LXical.Project{root_uri: "file:///Users/smithy/Repos/test", mix_exs_uri: "file:///Users/smithy/Repos/test/mix.exs", mix_project?: true, mix_env: nil, mix_target: nil, env_variables: %{}, project_module: nil, entropy: 11263})
Restart: :permanent
Shutdown: 5000
Type: :worker
22:19:48.061 [info] Child :erl_epmd of Supervisor :net_sup started
Pid: #PID<0.1557.0>
Start Call: :erl_epmd.start_link()
Restart: :permanent
Shutdown: 2000
Type: :worker
22:19:48.062 [info] Child :auth of Supervisor :net_sup started
Pid: #PID<0.1558.0>
Start Call: :auth.start_link()
Restart: :permanent
Shutdown: 2000
Type: :worker
22:19:48.064 [info] Child :net_kernel of Supervisor :net_sup started
Pid: #PID<0.1559.0>
Start Call: :net_kernel.start_link(%{name: :"manager-test-11263@127.0.0.1", supervisor: :net_sup_dynamic, name_domain: :longnames, clean_halt: false})
Restart: :permanent
Shutdown: 2000
Type: :worker
22:19:48.065 [debug] Child :net_sup_dynamic of Supervisor :kernel_sup started
Pid: #PID<0.1556.0>
Start Call: :erl_distribution.start_link(%{name: :"manager-test-11263@127.0.0.1", supervisor: :net_sup_dynamic, name_domain: :longnames, clean_halt: false})
Restart: :permanent
Shutdown: 1000
Type: :supervisor
22:19:51.069 [error] Process #PID<0.1555.0> terminating
** (exit) {:error, :start_timeout}
    (stdlib 6.2) gen_server.erl:2210: :gen_server.init_it/6
    (stdlib 6.2) proc_lib.erl:329: :proc_lib.init_p_do_apply/3
Initial Call: LXical.Server.Project.Node.init/1
Ancestors: [:"test::supervisor", LXical.Server.ProjectSupervisor, LXical.Server.Supervisor, #PID<0.1544.0>]
Message Queue Length: 0
Messages: []
Links: [#PID<0.1552.0>]
Dictionary: []
Trapping Exits: false
Status: :running
Heap Size: 1598
Stack Size: 29
Reductions: 224867
22:19:51.069 [error] Child {LXical.Server.Project.Node, "test"} of Supervisor :"test::supervisor" failed to start
** (exit) {:error, :start_timeout}
Start Call: LXical.Server.Project.Node.start_link(%LXical.Project{root_uri: "file:///Users/smithy/Repos/test", mix_exs_uri: "file:///Users/smithy/Repos/test/mix.exs", mix_project?: true, mix_env: nil, mix_target: nil, env_variables: %{}, project_module: nil, entropy: 11263})
Restart: :permanent
Shutdown: 5000
Type: :worker
22:19:51.075 [error] GenServer :"test::node_process" terminating
** (stop) :start_timeout
Last message: :maybe_start_timeout
State: nil
22:19:51.075 [error] Process :"test::node_process" (#PID<0.1562.0>) terminating
** (exit) :start_timeout
    (stdlib 6.2) gen_server.erl:2476: :gen_server.handle_common_reply/8
    (stdlib 6.2) proc_lib.erl:329: :proc_lib.init_p_do_apply/3
Initial Call: LXical.RemoteControl.ProjectNode.init/1
Ancestors: [LXical.RemoteControl.ProjectNodeSupervisor, :"test::supervisor", LXical.Server.ProjectSupervisor, LXical.Server.Supervisor, #PID<0.1544.0>]
Message Queue Length: 1
Messages: [{:EXIT, #PID<0.1554.0>, :shutdown}]
Links: [#Port<0.7>]
Dictionary: []
Trapping Exits: true
Status: :running
Heap Size: 10958
Stack Size: 29
Reductions: 29052
22:19:51.079 [error] Child :undefined of Supervisor LXical.RemoteControl.ProjectNodeSupervisor shut down abnormally
** (exit) :start_timeout
Pid: #PID<0.1562.0>
Start Call: LXical.RemoteControl.ProjectNode.start_link(%LXical.Project{root_uri: "file:///Users/smithy/Repos/test", mix_exs_uri: "file:///Users/smithy/Repos/test/mix.exs", mix_project?: true, mix_env: nil, mix_target: nil, env_variables: %{}, project_module: nil, entropy: 11263})
Restart: :transient
Shutdown: 5000
Type: :worker
22:19:51.081 [info] Lexical Initialized
22:19:51.081 [info] opened file:///Users/smithy/Repos/test/lib/test_web/router.ex
22:19:51.094 [info] Child of Supervisor :inet_gethost_native_sup started
Pid: #PID<0.1565.0>
Start Call: :inet_gethost_native.init([])
22:19:51.094 [debug] Child :inet_gethost_native_sup of Supervisor :kernel_safe_sup started
Pid: #PID<0.1564.0>
Start Call: :inet_gethost_native.start_link()
Restart: :temporary
Shutdown: 1000
Type: :worker
22:19:51.101 [error] ** (ErlangError) Erlang error: {:erpc, :noconnection}
    (kernel 10.2.2) erpc.erl:1376: :erpc.call/5
    (lx_server 0.7.2) lib/lexical/server/provider/handlers/document_symbols.ex:16: LXical.Server.Provider.Handlers.DocumentSymbols.handle/2
    (lx_server 0.7.2) lib/lexical/server/task_queue.ex:79: anonymous fn/4 in LXical.Server.TaskQueue.State.as_task/2
    (elixir 1.18.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.18.2) lib/task/supervised.ex:36: Task.Supervised.reply/4

22:19:51.101 [warning] Task Elixir.LXical.Server.Provider.Handlers.DocumentSymbols.handle/2 ran for  15 ms. Result :success
22:20:00.078 [error] ** (ErlangError) Erlang error: {:erpc, :noconnection}
    (kernel 10.2.2) erpc.erl:1376: :erpc.call/5
    (lx_server 0.7.2) lib/lexical/server/provider/handlers/document_symbols.ex:16: LXical.Server.Provider.Handlers.DocumentSymbols.handle/2
    (lx_server 0.7.2) lib/lexical/server/task_queue.ex:79: anonymous fn/4 in LXical.Server.TaskQueue.State.as_task/2
    (elixir 1.18.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.18.2) lib/task/supervised.ex:36: Task.Supervised.reply/4

22:20:00.078 [warning] Task Elixir.LXical.Server.Provider.Handlers.DocumentSymbols.handle/2 ran for  1 ms. Result :success
22:20:01.170 [error] ** (ErlangError) Erlang error: {:erpc, :noconnection}
    (kernel 10.2.2) erpc.erl:1376: :erpc.call/5
    (lx_server 0.7.2) lib/lexical/server/provider/handlers/document_symbols.ex:16: LXical.Server.Provider.Handlers.DocumentSymbols.handle/2
    (lx_server 0.7.2) lib/lexical/server/task_queue.ex:79: anonymous fn/4 in LXical.Server.TaskQueue.State.as_task/2
    (elixir 1.18.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.18.2) lib/task/supervised.ex:36: Task.Supervised.reply/4

22:20:01.171 [warning] Task Elixir.LXical.Server.Provider.Handlers.DocumentSymbols.handle/2 ran for  0.468 ms. Result :success
22:20:01.599 [error] ** (ErlangError) Erlang error: {:erpc, :noconnection}
    (kernel 10.2.2) erpc.erl:1376: :erpc.call/5
    (lx_server 0.7.2) lib/lexical/server/provider/handlers/document_symbols.ex:16: LXical.Server.Provider.Handlers.DocumentSymbols.handle/2
    (lx_server 0.7.2) lib/lexical/server/task_queue.ex:79: anonymous fn/4 in LXical.Server.TaskQueue.State.as_task/2
    (elixir 1.18.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.18.2) lib/task/supervised.ex:36: Task.Supervised.reply/4

22:20:01.599 [warning] Task Elixir.LXical.Server.Provider.Handlers.DocumentSymbols.handle/2 ran for  0.721 ms. Result :success
22:20:03.096 [error] ** (ErlangError) Erlang error: {:erpc, :noconnection}
    (kernel 10.2.2) erpc.erl:1376: :erpc.call/5
    (lx_server 0.7.2) lib/lexical/server/provider/handlers/document_symbols.ex:16: LXical.Server.Provider.Handlers.DocumentSymbols.handle/2
    (lx_server 0.7.2) lib/lexical/server/task_queue.ex:79: anonymous fn/4 in LXical.Server.TaskQueue.State.as_task/2
    (elixir 1.18.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.18.2) lib/task/supervised.ex:36: Task.Supervised.reply/4

22:20:03.096 [warning] Task Elixir.LXical.Server.Provider.Handlers.DocumentSymbols.handle/2 ran for  0.646 ms. Result :success
22:20:05.688 [error] ** (ErlangError) Erlang error: {:erpc, :noconnection}
    (kernel 10.2.2) erpc.erl:1376: :erpc.call/5
    (lx_server 0.7.2) lib/lexical/server/provider/handlers/document_symbols.ex:16: LXical.Server.Provider.Handlers.DocumentSymbols.handle/2
    (lx_server 0.7.2) lib/lexical/server/task_queue.ex:79: anonymous fn/4 in LXical.Server.TaskQueue.State.as_task/2
    (elixir 1.18.2) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.18.2) lib/task/supervised.ex:36: Task.Supervised.reply/4

22:20:05.688 [warning] Task Elixir.LXical.Server.Provider.Handlers.DocumentSymbols.handle/2 ran for  0.958 ms. Result :success
```

rdlu commented:
> > > Is this kinda dead in the water? I know they're working on the unified LS - but is lexical still receiving updates in the meantime?
> > 
> > 
> > I was waiting this pr to be merged as well, but then I installed manually using the 1.18 branch. I went back to next-ls for a while, but this is much faster on my setup.
> > Cloning, switching to `support_1_18` branch and then `mix package --path ~/.local/bin/lexical` was less painful than I imagined.
> > It works with my lazyvim/neovim setup: https://github.com/rdlu/dotfiles/blob/main/lazyvim/.config/nvim/lua/plugins/elixir.lua
> > Also with zed editor I just symlinked to `~/.local/share/zed/extensions/work/elixir/lexical-v0.7.x`
> 
> hmm I tried your config (I'm also using lazyvim so there wasn't really much difference except the build path) and built from latest commit on this branch, but it wouldn't run so seems some issues still. FWIW, I made sure to build on the same version of elixir/otp as I ran in the test repo. elixir 1.18.2 and otp 27.2.2
> 
> My lexical.log attached. [test.log](https://github.com/user-attachments/files/18707214/test.log)

By looking at you log you are using Windows, am I right?

I'm not sure if I can help, I'm exclusively using Linux since eons ago (sometimes macos)

But sometimes my neovim or zed editor get stuck, the suggestions stop appearing, `gd` stops working, etc.

My workaround to avoid this is whenever there are major changes to versions I prefer to clean everything that was built. I have this in my `justfile`
that I use for my projects, then I `mix deps.get` again to start from a clean build state:

```
clean-dev:
  rm -rf ./.elixir_ls ./_build ./deps ./.elixir-tools
  find . -name "*.beam" -exec rm -f {} \;
```

garlandcrow commented:
> > > > Is this kinda dead in the water? I know they're working on the unified LS - but is lexical still receiving updates in the meantime?
> > > 
> > > 
> > > I was waiting this pr to be merged as well, but then I installed manually using the 1.18 branch. I went back to next-ls for a while, but this is much faster on my setup.
> > > Cloning, switching to `support_1_18` branch and then `mix package --path ~/.local/bin/lexical` was less painful than I imagined.
> > > It works with my lazyvim/neovim setup: https://github.com/rdlu/dotfiles/blob/main/lazyvim/.config/nvim/lua/plugins/elixir.lua
> > > Also with zed editor I just symlinked to `~/.local/share/zed/extensions/work/elixir/lexical-v0.7.x`
> > 
> > 
> > hmm I tried your config (I'm also using lazyvim so there wasn't really much difference except the build path) and built from latest commit on this branch, but it wouldn't run so seems some issues still. FWIW, I made sure to build on the same version of elixir/otp as I ran in the test repo. elixir 1.18.2 and otp 27.2.2
> > My lexical.log attached. [test.log](https://github.com/user-attachments/files/18707214/test.log)
> 
> By looking at you log you are using Windows, am I right?
> 
> I'm not sure if I can help, I'm exclusively using Linux since eons ago (sometimes macos)
> 
> But sometimes my neovim or zed editor get stuck, the suggestions stop appearing, `gd` stops working, etc.
> 
> My workaround to avoid this is whenever there are major changes to versions I prefer to clean everything that was built. I have this in my `justfile` that I use for my projects, then I `mix deps.get` again to start from a clean build state:
> 
> ```
> clean-dev:
>   rm -rf ./.elixir_ls ./_build ./deps ./.elixir-tools
>   find . -name "*.beam" -exec rm -f {} \;
> ```

No worries, was just posted in case this helps a maintainer, I'm more than used to elixir tools being broken and unusable, so just check in from time to time to see if they work in neovim (spoiler, they never do). I'm running macos, super standard nothing special, default everything with neovim/lazynvim. Nothing ever works, its cool ✌🏼

scohen commented:
People, i'm going to work on getting this merged in the next two days. My day job is taking up a lot of my time.

scohen committed:
removed 1.13 from testing matrix

scohen committed:
[feat] Massively sped up versions checks

The version of elixir and erlang won't change while the VM is running,
so they're a perfect example of what should be cached in
persistent_term.

scohen committed:
Get all tests to pass

* Skip tests broken by elixir sense update
* Update stream data
* remove compile warning

scohen committed:
Mise, moved its releases and broke the world... again

scohen committed:
trying again

scohen committed:
tweaked architecture to include OS

scohen committed:
adding verbose flag

scohen committed:
removing elixir install as it's a core plugin

scohen committed:
asdf update is not supported

scohen committed:
re-added 1.13

garlandcrow commented:
thanks for merging something that only works on your machine, greatly appreciated

rverton commented:
Works great here, thanks a lot for your efforts!
